### PR TITLE
Skip device authz when issuing App or Windows certs

### PR DESCRIPTION
**Pull request by codingllama**
Skip device authorization for App or WindowsDesktop certificates - those modes don't yet support device trust, so there is no point in requiring a trusted device.

This fixes an issue for clusters that set the global `device_trust.mode` to `required` and have App or WindowsDesktop servers.

https://github.com/gravitational/teleport.e/issues/514

### CODE REVIEW

**Comment by codingllama** [comment]
FYI @zmb3 @r0mant - I don't think we need to hold v12 for this one, but it would be nice to have it. If you want to wait, feel free to add the label.

File: lib/auth/grpcserver.go [hunk 1]
```diff
@@ -2500,6 +2498,8 @@ func validateUserSingleUseCertRequest(ctx context.Context, actx *grpcContext, re
 		}
 	case proto.UserCertsRequest_All:
 		return trace.BadParameter("must specify a concrete Usage in UserCertsRequest, one of SSH, Kubernetes or Database")
+	case proto.UserCertsRequest_App:
```
Comment (zmb3) [inline lib/auth/grpcserver.go:2501]: this change is unrelated to device trust, right?

Comment (codingllama) [reply lib/auth/grpcserver.go:2501]: Correct, it's just a custom error message, otherwise it falls to the default of "unknown certificate Usage".

**Comment by codingllama** [comment]
Friendly ping @fspmarshall @rudream ?

**Comment by codingllama** [comment]
Friendly ping @fspmarshall @rudream ?

**Review by zmb3 — APPROVED** [review]

File: lib/auth/auth_with_roles.go [hunk 2]
```diff
@@ -2778,6 +2774,23 @@ func (a *ServerWithRoles) generateUserCerts(ctx context.Context, req proto.UserC
...
 }
 
+// verifyUserDeviceForCertIssuance verifies if the user device is a trusted
+// device, in accordance to the certificate usage and the cluster's DeviceTrust
+// settings. It's meant to be called before issuing new user certificates.
+// Each Node (or access server) verifies the device independently, so this check
+// is not paramount to the access system itself, but it stops bad attempts from
+// progressing further and provides better feedback than other protocol-specific
+// failures.
+func (a *ServerWithRoles) verifyUserDeviceForCertIssuance(usage proto.UserCertsRequest_CertUsage, dt *types.DeviceTrust) error {
```
Comment (tobiaszheller) [inline lib/auth/auth_with_roles.go:2784]: nit: maybe this fn can include:
```
	authPref, err := a.authServer.GetAuthPreference(ctx)
	if err != nil {
		return nil, trace.Wrap(err)
```

**Review by tobiaszheller — APPROVED** [review]

**Review by rosstimothy — APPROVED** [review]

File: lib/auth/auth_with_roles.go (hunk 2 above)
Comment (codingllama) [reply lib/auth/auth_with_roles.go:2784]: Ack. I think I'd rather keep that outside, as callers could arguably use other info from the authPref themselves, plus we keep the parameters more targeted.

**Comment by codingllama** [comment]
Thanks for the quick review, folks.
